lib/src/rules/unnecessary_escapes_in_string.dart

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:meta/meta.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Remove unnecessary backslashes in strings.';
+
+const _details = r'''
+
+Remove unnecessary backslashes in strings.
+
+**BAD:**
+```
+'this string contains 2 \"double quotes\" ';
+"this string contains 2 \'single quotes\' ";
+```
+
+**GOOD:**
+```
+'this string contains 2 "double quotes" ';
+"this string contains 2 'single quotes' ";
+```
+
+''';
+
+class UnnecessaryEscapesInString extends LintRule implements NodeLintRule {
+  UnnecessaryEscapesInString()
+      : super(
+            name: 'unnecessary_escapes_in_string',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this);
+    registry.addSimpleStringLiteral(this, visitor);
+    registry.addStringInterpolation(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitSimpleStringLiteral(SimpleStringLiteral node) {
+    if (node.isRaw) return;
+
+    visitLexeme(
+      node.literal,
+      isSingleQuoted: node.isSingleQuoted,
+      isMultiline: node.isMultiline,
+      contentsOffset: node.contentsOffset,
+      contentsEnd: node.contentsEnd,
+    );
+  }
+
+  @override
+  void visitStringInterpolation(StringInterpolation node) {
+    for (var element in node.elements.whereType<InterpolationString>()) {
+      visitLexeme(
+        element.contents,
+        isSingleQuoted: node.isSingleQuoted,
+        isMultiline: node.isMultiline,
+        // TODO(a14n): should be the following line but the values look buggy
+        // contentsOffset: element.contentsOffset,
+        // contentsEnd: element.contentsEnd,
```

Comment:
I faced strange behaviour with the above code. It looks like a bug to me in the returned values.

For instance `"\'$f"`:

```
node:"\'$f"
node.offset:286
element:"\'
element.offset:286
element.contentsOffset:287   // OK
element.contentsEnd:288      // KO should be 289
element.end:289              // OK
element:"
element.offset:291           // OK
element.contentsOffset:292   // KO should be 291
element.contentsEnd:292      // KO should be 291
element.end:292              // OK
```

@bwilkerson WDYT?

Reply:
I agree, those values look wrong.

Reply:
@a14n: maybe open a bug on analyzer to track?  (fyi @stereotype441)

Reply:
Filed dart-lang/sdk#40495